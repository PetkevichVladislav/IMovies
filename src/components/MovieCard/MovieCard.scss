.movie-card{
    max-width: 322px;
    margin-top: 25px;
    position: relative;

    &__description-container{
        display: flex;
        justify-content: space-between;
        margin: 22px 0px 0px 0px;
        align-items: center;
    }

    &__description{
        margin: 0px;
        color: #FFFFFF;

        &--title{
            font-size: 18px;
        }

        &--production-year{
            border: #97979786 1px solid;
            border-radius: 4px;
            padding: 5px 8px;
        }

        &--genre{
            margin: 8px 0px 0px 0px;
            opacity: 50%;
        }
    }

    &__context-menu{
        position: absolute;
        display: inline-block;
        margin-left: auto;
        background-color: #2A202D;
        border-radius: 50%;
        width: 36px;
        height: 36px;
        top:15px;
        right: 15px;
    }

    &__popupmenu_container {
        position: absolute;
        top:15px;
        right: 15px;
    }

    &__image{
        width: 322px;
        height: 455px;
    }

    &__image-container{
        background-color: #FFFFFF;
        margin: 0px;
        width: 322px;
        height: 455px;
        cursor: pointer;
    }
}

.icon-more-vertical::before{
    content: "\e905";
    color: white;
    display: block;
    font-size: 32px;
    padding-top: 1px;
    padding-left: 2px;
}