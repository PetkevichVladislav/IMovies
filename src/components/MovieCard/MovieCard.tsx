import "./MovieCard.scss"

import { FC, useMemo, useState } from "react";
import { PopupMenu } from "../PopupMenu/PopupMenu";
import { CardModel } from "../../models/CardModel";

export interface MenuModel {
    onEditCLick: () => void;
    onDeleteClick: () => void;
}

export interface IMovieCard {
    card : CardModel;
    id: string;
    onCardClick: () => void;
    movieMenu: MenuModel;
}

export const MovieCard: FC<IMovieCard> = ({card, id, onCardClick, movieMenu}) => {
    const [isMouseHoveredOnCard, setIsMouseHoveredOnCard] = useState<boolean>(false);
    const [isPopupMenuOpened, setIsPopupMenuOpened] = useState<boolean>(false);
    const menuItems = useMemo(() => [
        {
            onClick: movieMenu.onEditCLick,
            menuItem: 'Edit',
        },
        {
            onClick: movieMenu.onDeleteClick,
            menuItem: 'Delete',
        }
    ], [movieMenu]);

    const handleClosePopupMenu = () => {
        setIsPopupMenuOpened(false);
    };

    const handleMouseEnter = () => {
        setIsMouseHoveredOnCard(true);
    }

    const handleMouseLeave = () => {
        setIsPopupMenuOpened(false);
        setIsMouseHoveredOnCard(false);
    }

    const handleCardClick = () => {
        onCardClick();
    }

    const handleMenuClick = (event: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
        event.stopPropagation();
        setIsPopupMenuOpened(true);
    }

    return (
        <div className="movie-card" id={id} data-testid="movie-card" onMouseEnter={handleMouseEnter} onMouseLeave={handleMouseLeave} onClick={handleCardClick}>
            {isMouseHoveredOnCard &&
                <span className="movie-card__context-menu icon-more-vertical" data-testid="card-menu-item" onClick={(event) => handleMenuClick(event)}></span>
            }
            {isPopupMenuOpened &&
                <div className="movie-card__popupmenu_container">
                    <PopupMenu menuItems={menuItems} onClose={handleClosePopupMenu} />
                </div>
            }
            <figure className="movie-card__image-container">
            <img className="movie-card__image" src={card.imageUrl} alt={card.movieName + "_movie_poster"} />
            </figure>
            <div className="movie-card__description-container">
                <p className="movie-card__description movie-card__description--title">{card.movieName}</p>
                <p className="movie-card__description movie-card__description--production-year">{card.releaseYear}</p>
            </div>
            <p className="movie-card__description movie-card__description--genre">{card.genres.join(", ")}</p>
        </div>
    );
}