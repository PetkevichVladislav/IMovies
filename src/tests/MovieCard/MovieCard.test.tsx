import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { IMovieCard, MovieCard } from '../../components/MovieCard/MovieCard';

describe("Movie card", () => {
    const onClickCallback = jest.fn();
    const onRemoveMenuCallback = jest.fn();
    const onEditMenuCallback = jest.fn();
    const movieData: IMovieCard = {
        id: "1",
        card: {
            imageUrl: "http://uimage.com",
            movieName: 'Example Movie',
            releaseYear: 2022,
            genres: ['Action', 'Drama'],
        },
        onCardClick: onClickCallback,
        movieMenu: {
            onDeleteClick: onRemoveMenuCallback,
            onEditCLick: onEditMenuCallback,
        }
    };

    it("When data is valid component should renders correctly", () => {
        const tree = renderer.create(<MovieCard {...movieData} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("when user open forms the data should reneders data", () => {
        render(<MovieCard {...movieData} />);
        expect(screen.getByRole("img")).toHaveAttribute("src", "http://uimage.com");
        expect(screen.getByText('Example Movie')).toBeInTheDocument();
        expect(screen.getByText('2022')).toBeInTheDocument();
        expect(screen.getByText('Action, Drama')).toBeInTheDocument();
    });

    it("when user clicks on element callback should be triggered", () => {
        render(<MovieCard {...movieData} />);
        const card = screen.getByTestId("movie-card");
        fireEvent.click(card);
        expect(onClickCallback).toBeCalled();
    });

    it("when user hover on card menu item button should appear", () => {
        render(<MovieCard {...movieData} />);
        const card = screen.getByTestId("movie-card");
        fireEvent.mouseEnter(card);
        const menuItem = screen.getByTestId("card-menu-item");
        expect(menuItem).toBeInTheDocument();
    });

    it("when user clicks on card menu item button menu should appear", () => {
        render(<MovieCard {...movieData} />);
        const card = screen.getByTestId("movie-card");
        fireEvent.mouseEnter(card);
        const menuItem = screen.getByTestId("card-menu-item");
        fireEvent.click(menuItem);
        const menu = screen.getByTestId("drop-down-menu");
        expect(menu).toBeInTheDocument();
    });

});