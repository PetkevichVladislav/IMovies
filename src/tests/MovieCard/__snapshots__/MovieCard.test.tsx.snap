// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie card When data is valid component should renders correctly 1`] = `
<div
  className="movie-card"
  data-testid="movie-card"
  id="1"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <figure
    className="movie-card__image-container"
  >
    <img
      alt="Example Movie_movie_poster"
      className="movie-card__image"
      src="http://uimage.com"
    />
  </figure>
  <div
    className="movie-card__description-container"
  >
    <p
      className="movie-card__description movie-card__description--title"
    >
      Example Movie
    </p>
    <p
      className="movie-card__description movie-card__description--production-year"
    >
      2022
    </p>
  </div>
  <p
    className="movie-card__description movie-card__description--genre"
  >
    Action, Drama
  </p>
</div>
`;
